### DISCUSSION TITLE
Create simple and custom tooltip widget

### DISCUSSION THREAD

**Discussion by dfguerrerom** [12rambau/sepal_ui#168] [Show and tell]
Many times I have had the need of have a tooltip for some widgets, the current implementation of ipyvuetify could cause some confusion when implementing, and in one of my projects I just created a simple way, so I guess it could be widely implemented in sepal_ui, what do you think?

```Python
class Tooltip(v.Tooltip):
    
    def __init__(self, widget, tooltip, *args, **kwargs):
        """
        Custom widget to display tooltip when mouse is over widget

        Args:
            widget (DOM.widget): widget used to display tooltip
            tooltip (str): the text to display in the tooltip
            
        Example:
            
            btn = v.Btn(children=['Button'])
            Tooltip(widget=btn, tooltip='Click over the button')
        """
        
        self.bottom=True
        self.v_slots=[{
            'name': 'activator',
            'variable': 'tooltip',
            'children': widget
        }]
        widget.v_on = 'tooltip.on'
        
        self.children = [tooltip]
        
        super().__init__(*args, **kwargs)
```

**Comment by 12rambau** [comment]
I love it because I agree that slots are a nightmare

is `tooltip` an available slot for all widgets ? because if not you'll need to add some sort of security if you don't want to see everything disapear 😄

**Reply by dfguerrerom** [reply]
Yes, it will be available for all widgets... however... if you'd like to change any of the traits of the tooltip, your widget will disappear...
but, as tooltips traits are not such susceptible to change, I guess there won't be any problem... for example:
```Python

# we can create a button widget
btn = v.Btn(children=['button'])

# wrap it with the tooltip
btn_tool = Tooltip(widget=btn, tooltip='This is a tooltip')

# and display it 
btn_tool 

# what we can always do is modify the widget traits without any problems
btn.children=['new button name']

# but if we'd like to change any trait of the btn_tool, our widget will disappear 
# (at least while the ipyvuetify slot bug remains)
btn_tool .children=['new button name']

```

what do you think?

**Reply by 12rambau** [reply]
I think you can prevent the changes to an object after its init. I know how to do it in C so I'm sure there is a trick in Python

**Reply by 12rambau** [reply]
```python
class Immutable(object):
    def __setattr__(self, key, value):
        if not hasattr(self, key):
            super().__setattr__(key, value)
        else:
            raise RuntimeError("Can't modify immutable object's attribute: {}".format(key))
```

This solution allows to create a fully immutable object, I think it's the best way to go

**Comment by 12rambau** [comment]
You can move it an issue and push it to develop. 
I'll need to add the copy-to-clip boar And I'll push a new version
